qlib/contrib/strategy/cost_control.py

```diff
@@ -89,13 +85,15 @@ def generate_target_weight_position(self, score, current, trade_start_time, trad
                         max(1 / self.topk - final_stock_weight.get(stock_id, 0), 0.0),
                         sold_stock_weight,
                     )
-                    final_stock_weight[stock_id] = final_stock_weight.get(stock_id, 0.0) + add_weight
+                    final_stock_weight[stock_id] = (
+                        final_stock_weight.get(stock_id, 0.0) + add_weight
+                    )
                     sold_stock_weight -= add_weight
             elif self.buy_method == "average_fill":
                 for stock_id in buy_signal_stocks:
-                    final_stock_weight[stock_id] = final_stock_weight.get(stock_id, 0.0) + sold_stock_weight / len(
-                        buy_signal_stocks
-                    )
+                    final_stock_weight[stock_id] = final_stock_weight.get(
+                        stock_id, 0.0
+                    ) + sold_stock_weight / len(buy_signal_stocks)
             else:
```

Comment:
The changes made here did not involve any modifications to the calculation logic, and the new format did not pass the CI testing. Can we revert back to the previous format?

Reply:
Yes, but lint is another problem here and I am confused by conflict between ci and pre-commit, no flake8 in ci and it is not actually used in code. So I can not enable pre-commit in my develop env.

https://github.com/microsoft/qlib/blob/05d67b38287a22d43d55aee76e8fb3bba87c5f48/.github/workflows/test_qlib_from_source.yml#L71

https://github.com/microsoft/qlib/blob/05d67b38287a22d43d55aee76e8fb3bba87c5f48/.pre-commit-config.yaml#L11-L12

```
pre-commit run --all-files
black....................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

examples/data_demo/data_mem_resuse_demo.py:10:1: F401 'pickle' imported but unused
examples/data_demo/data_mem_resuse_demo.py:12:1: F401 'subprocess' imported but unused
examples/online_srv/online_management_simulate.py:11:1: F401 'qlib.model.trainer.DelayTrainerR' imported but unused
examples/online_srv/online_management_simulate.py:11:1: F401 'qlib.model.trainer.DelayTrainerRM' imported but unused
examples/online_srv/online_management_simulate.py:103:9: F841 local variable 'CSI300_BENCH' is assigned to but never used
tests/rl/test_saoe_simple.py:20:1: F403 'from qlib.rl.order_execution import *' used; unable to detect undefined names
tests/rl/test_saoe_simple.py:51:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:85:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:108:21: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:111:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:121:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:140:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:148:19: F405 'FullHistoryStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:149:24: F405 'CurrentStepStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:150:26: F405 'CategoricalActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:151:31: F405 'TwapRelativeActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:221:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:227:19: F405 'FullHistoryStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:228:21: F405 'CategoricalActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:232:15: F405 'Recurrent' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:233:14: F405 'PPO' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:255:20: F405 'FullHistoryStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:256:21: F405 'TwapRelativeActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:257:14: F405 'AllOne' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:261:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:284:20: F405 'FullHistoryStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:285:21: F405 'CategoricalActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:286:15: F405 'Recurrent' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:287:14: F405 'PPO' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:292:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:315:20: F405 'FullHistoryStateInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:316:21: F405 'CategoricalActionInterpreter' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:317:15: F405 'Recurrent' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:318:14: F405 'PPO' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:321:17: F405 'SingleAssetOrderExecutionSimple' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/rl/test_saoe_simple.py:326:9: F405 'PAPenaltyReward' may be undefined, or defined from star imports: qlib.rl.order_execution
tests/test_contrib_model.py:14:17: F841 local variable 'model' is assigned to but never used
examples/benchmarks/TFT/tft.py:219:9: F841 local variable 'best_loss' is assigned to but never used
examples/benchmarks/TFT/tft.py:258:9: F841 local variable 'use_gpu' is assigned to but never used
examples/benchmarks/TFT/tft.py:273:13: F841 local variable 'targets' is assigned to but never used
tests/rolling_tests/test_update_pred.py:5:1: F401 'fire' imported but unused
tests/rolling_tests/test_update_pred.py:8:1: F401 'qlib' imported but unused
tests/rolling_tests/test_update_pred.py:50:9: F841 local variable 'pred' is assigned to but never used
tests/rolling_tests/test_update_pred.py:113:9: F841 local variable 'pred' is assigned to but never used
scripts/data_collector/cn_index/collector.py:7:1: F401 'datetime' imported but unused
tests/data_mid_layer_tests/test_handler_storage.py:2:1: F401 'time' imported but unused
examples/benchmarks/TFT/libs/utils.py:36:5: E741 ambiguous variable name 'l'
scripts/data_collector/yahoo/collector.py:5:1: F401 're.I' imported but unused
scripts/data_collector/yahoo/collector.py:114:9: F841 local variable 'e' is assigned to but never used
scripts/data_collector/yahoo/collector.py:144:9: F841 local variable 'e' is assigned to but never used
scripts/data_collector/yahoo/collector.py:173:13: F841 local variable 'e' is assigned to but never used
scripts/data_collector/yahoo/collector.py:183:17: F841 local variable 'e' is assigned to but never used
scripts/data_collector/fund/collector.py:92:9: F841 local variable 'e' is assigned to but never used
scripts/data_collector/us_index/collector.py:7:1: F401 'importlib' imported but unused
examples/benchmarks_dynamic/DDG-DA/vis_data.py:2:1: F401 'numpy as np' imported but unused
examples/benchmarks_dynamic/DDG-DA/vis_data.py:10:1: F401 'tqdm.auto.tqdm' imported but unused
examples/benchmarks/TRA/src/model.py:19:1: F401 'qlib.utils.get_or_create_path' imported but unused
examples/orderbook_data/create_dataset.py:9:1: F401 'datetime.date' imported but unused
examples/orderbook_data/create_dataset.py:9:1: F401 'datetime.datetime as dt' imported but unused
examples/orderbook_data/create_dataset.py:12:1: F401 'random' imported but unused
examples/orderbook_data/create_dataset.py:17:1: F401 'arctic.chunkstore' imported but unused
examples/orderbook_data/create_dataset.py:18:1: F401 'arctic' imported but unused
examples/orderbook_data/create_dataset.py:19:1: F811 redefinition of unused 'Arctic' from line 17
examples/orderbook_data/create_dataset.py:20:1: F401 'arctic.chunkstore.chunkstore.CHUNK_SIZE' imported but unused
examples/orderbook_data/create_dataset.py:22:1: F401 'joblib.parallel' imported but unused
examples/orderbook_data/create_dataset.py:23:1: F401 'numpy as np' imported but unused
examples/orderbook_data/create_dataset.py:25:1: F401 'pandas.DataFrame' imported but unused
examples/orderbook_data/create_dataset.py:26:1: F401 'pandas.core.indexes.datetimes.date_range' imported but unused
examples/orderbook_data/create_dataset.py:59:67: F811 redefinition of unused 'date' from line 9
examples/orderbook_data/create_dataset.py:94:45: E262 inline comment should start with '# '
examples/orderbook_data/create_dataset.py:102:64: E262 inline comment should start with '# '
examples/orderbook_data/create_dataset.py:127:21: E712 comparison to True should be 'if cond is True:' or 'if cond:'
examples/orderbook_data/create_dataset.py:136:77: F811 redefinition of unused 'date' from line 9
examples/orderbook_data/create_dataset.py:278:40: F811 redefinition of unused 'date' from line 9
examples/orderbook_data/create_dataset.py:310:72: F811 redefinition of unused 'date' from line 9
scripts/data_collector/br_index/collector.py:6:1: F401 'importlib' imported but unused
examples/highfreq/highfreq_ops.py:1:1: F401 'numpy as np' imported but unused
examples/highfreq/highfreq_ops.py:3:1: F401 'importlib' imported but unused
examples/highfreq/highfreq_ops.py:5:1: F401 'qlib.config.C' imported but unused
examples/highfreq/highfreq_ops.py:6:1: F401 'qlib.data.cache.H' imported but unused
examples/highfreq/highfreq_ops.py:7:1: F401 'qlib.data.data.Cal' imported but unused
examples/highfreq/highfreq_ops.py:149:33: E741 ambiguous variable name 'l'
examples/highfreq/highfreq_ops.py:150:14: E741 ambiguous variable name 'l'
examples/online_srv/rolling_online_management.py:16:1: F401 'qlib.model.trainer.DelayTrainerR' imported but unused
examples/online_srv/rolling_online_management.py:16:1: F401 'qlib.model.trainer.TrainerR' imported but unused
examples/online_srv/rolling_online_management.py:16:1: F401 'qlib.model.trainer.end_task_train' imported but unused
examples/online_srv/rolling_online_management.py:16:1: F401 'qlib.model.trainer.task_train' imported but unused
tests/data_mid_layer_tests/test_dataset.py:6:1: F401 'sys' imported but unused
tests/test_contrib_workflow.py:70:9: F841 local variable 'uri_path' is assigned to but never used
tests/test_contrib_workflow.py:74:9: F841 local variable 'uri_path' is assigned to but never used
tests/test_pit.py:11:1: F401 'baostock as bs' imported but unused
scripts/dump_pit.py:9:1: F401 'abc' imported but unused
scripts/dump_pit.py:12:1: F401 'traceback' imported but unused
scripts/dump_pit.py:14:1: F401 'typing.List' imported but unused
scripts/dump_pit.py:14:1: F401 'typing.Union' imported but unused
scripts/dump_pit.py:16:1: F401 'concurrent.futures.ThreadPoolExecutor' imported but unused
scripts/dump_pit.py:19:1: F401 'numpy as np' imported but unused
scripts/dump_pit.py:23:1: F401 'qlib.utils.code_to_fname' imported but unused
scripts/data_collector/utils.py:85:17: F811 redefinition of unused '_get_calendar' from line 70
scripts/data_collector/utils.py:206:9: F841 local variable '_retry' is assigned to but never used
tests/rl/test_data_queue.py:64:60: F841 local variable 'data_queue' is assigned to but never used
tests/rl/test_data_queue.py:78:66: F841 local variable 'data_queue' is assigned to but never used
tests/data_mid_layer_tests/test_handler.py:3:1: F401 'shutil' imported but unused
examples/workflow_by_code.py:4:39: W291 trailing whitespace
examples/highfreq/workflow.py:13:1: F401 'qlib.data.ops.Operators' imported but unused
examples/highfreq/workflow.py:122:9: E265 block comment should start with '# '
examples/highfreq/workflow.py:127:9: E265 block comment should start with '# '
examples/highfreq/workflow.py:135:9: E265 block comment should start with '# '
examples/highfreq/workflow.py:167:9: E265 block comment should start with '# '
scripts/data_collector/crypto/collector.py:8:1: F401 'requests' imported but unused
scripts/data_collector/crypto/collector.py:41:9: E722 do not use bare 'except'
scripts/data_collector/crypto/collector.py:124:9: F841 local variable 'e' is assigned to but never used
scripts/data_collector/base.py:11:1: F401 'concurrent.futures.ThreadPoolExecutor' imported but unused
scripts/data_collector/base.py:86:13: F841 local variable 'e' is assigned to but never used
tests/backtest/test_high_freq_trading.py:1:1: F401 'typing.List' imported but unused
tests/backtest/test_high_freq_trading.py:1:1: F401 'typing.Tuple' imported but unused
tests/backtest/test_high_freq_trading.py:1:1: F401 'typing.Union' imported but unused
tests/backtest/test_high_freq_trading.py:2:1: F401 'qlib.backtest.position.Position' imported but unused
tests/backtest/test_high_freq_trading.py:4:1: F401 'qlib.backtest.decision.BaseTradeDecision' imported but unused
tests/backtest/test_high_freq_trading.py:5:1: F401 'qlib' imported but unused
tests/backtest/test_high_freq_trading.py:124:9: F841 local variable 'report' is assigned to but never used
tests/backtest/test_high_freq_trading.py:127:9: F841 local variable 'f_dec' is assigned to but never used
tests/test_all_pipeline.py:4:1: F401 'sys' imported but unused
tests/test_all_pipeline.py:10:1: F401 'qlib' imported but unused
tests/test_all_pipeline.py:11:1: F401 'qlib.config.C' imported but unused
examples/benchmarks_dynamic/baseline/rolling_benchmark.py:11:1: F401 'tqdm.auto.tqdm' imported but unused
docs/conf.py:21:1: F401 'os' imported but unused
docs/conf.py:22:1: F401 'sys' imported but unused
examples/nested_decision_execution/workflow.py:105:1: F401 'qlib.data.D' imported but unused
examples/nested_decision_execution/workflow.py:106:1: F401 'qlib.utils.exists_qlib_data' imported but unused
examples/data_demo/data_cache_demo.py:10:1: F401 'pickle' imported but unused
examples/orderbook_data/example.py:4:1: F401 'arctic.arctic.Arctic' imported but unused
examples/rolling_process_data/rolling_handler.py:2:1: F401 'qlib.data.dataset.loader.DataLoaderDH' imported but unused
```

Reply:
I write a Makefile to support the pre-commit check.
Like run 'make precommit' before I commit the code.
 
Makefile:
`
.PHONY: precommit
precommit:  ## Lint and static-check
	black . -l 120
	pylint --disable=C0104,C0114,C0115,C0116,C0301,C0302,C0411,C0413,C1802,R0401,R0801,R0902,R0903,R0911,R0912,R0913,R0914,R0915,R1720,W0105,W0123,W0201,W0511,W0613,W1113,W1514,E0401,E1121,C0103,C0209,R0402,R1705,R1710,R1725,R1735,W0102,W0212,W0221,W0223,W0231,W0237,W0612,W0621,W0622,W0703,W1309,E1102,E1136 --const-rgx='[a-z_][a-z0-9_]{2,30}$' qlib --init-hook "import astroid; astroid.context.InferenceContext.max_inferred = 500; import sys; sys.setrecursionlimit(2000)"

        flake8 --ignore=E501,F541,E266,E402,W503,E731,E203 --per-file-ignores="__init__.py:F401,F403" qlib
`

> Yes, but lint is another problem here and I am confused by conflict between ci and pre-commit, no flake8 in ci and it is not actually used in code. So I can not enable pre-commit in my develop env.
>